irohad/ametsuchi/impl/flat_file/flat_file.cpp

```diff
@@ -16,145 +16,221 @@
+        // with error "nudb::error::key_not_found". If we did not get this type
+        // of error -- something bad happened.
+        log_->critical("can not read database to count blocks: {}",
+                       ec.message());
+        return boost::none;
+      }
+
+      auto last_id = total_blocks == 0u ? 0u : total_blocks - 1;
+
+      return std::unique_ptr<FlatFile>(
+          new FlatFile(std::move(db), path, last_id));
+    }
+
+    const std::string &FlatFile::directory() const {
+      return path_;
+    }
+
+    bool FlatFile::add(Identifier id, const std::vector<uint8_t> &blob) {
+      nudb::error_code ec;
+      auto key = serialize_uint32(id);
+      db_->insert(key.data(), blob.data(), blob.size(), ec);
+      if (ec) {
+        log_->error("add(): {}", ec.message());
+        return false;
+      }
+      current_id_.store(id);
+      return true;
+    }
+
+    boost::optional<std::vector<uint8_t>> FlatFile::get(Identifier id) const {
+      nudb::error_code ec;
+      boost::optional<std::vector<uint8_t>> ret;
+      auto key = serialize_uint32(id);
+      db_->fetch(key.data(),
+                 [&ret](const void *p, size_t size) {
+                   if (size == 0) {
+                     ret = boost::none;
+                   } else {
+                     const auto *c = static_cast<const uint8_t *>(p);
+                     ret = std::vector<uint8_t>{c, c + size};
+                   }
+                 },
+                 ec);
+      if (ec) {
+        log_->error("get(): {}", ec.message());
+        return boost::none;
+      }
+
+      return ret;
+    }
+
+    bool FlatFile::dropAll() {
+      nudb::error_code ec1, ec2, ec3;
+
+      nudb::erase_file(db_->dat_path(), ec1);
+      nudb::erase_file(db_->log_path(), ec2);
+      nudb::erase_file(db_->key_path(), ec3);
+
+      current_id_.store(0);
+
+      return not(ec1 or ec2 or ec3);
+    }
+
+    uint32_t FlatFile::count_blocks(nudb::store &db, nudb::error_code &ec) {
+      auto log_ = logger::log("FlatFile::count_blocks");
+
+      FlatFile::Identifier current = FIRST_BLOCK_AT;
+
+      bool found_last = false;
+
+      do {
+        auto key = serialize_uint32(current);
+
+        db.fetch(key.data(),
+                 [&found_last, &current](const void *value, size_t size) {
+                   // if we read 0 bytes, then there is no such key
+                   if (size == 0u) {
+                     found_last = true;
+                     return;
+                   }
+
+                   // go to the next key
+                   current++;
+                 },
+                 ec);
+
+        if (ec == nudb::error::key_not_found) {
+          return current;
+        } else if (ec) {
+          // some other error occurred
+          log_->error("{}", ec.message());
+          return 0;
+        }
+      } while (not found_last);
+
+      return current;
+    }
+
+    std::array<uint8_t, sizeof(uint32_t)> FlatFile::serialize_uint32(
```

Comment:
It is bad practice when you talk about numbers, because of implicit conversions. I explicitly want to serialize 4 bytes.
Anyway, it is an implementation detail, which is not very important.